win_wait_for_process: Fixes and integration tests

dagwieers opened the pull request:
##### SUMMARY
Refactor code and add integration tests to win_wait_for_process

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
win_wait_for_process

##### ANSIBLE VERSION
v2.7

dagwieers committed:
win_wait_for_process: Add integration tests

dagwieers committed:
Disable reporting changes

dagwieers committed:
Added more tests checking PID

dagwieers commented:
While creating these integration tests, 2 issues popped up:
- Some of the failure messages are incorrect (as `$process_name` is not set)
- The module does not always return `matched_processes`

A few other issues were dealt with:
- Module returning changes in some occasions (module does not make changes !)
- Module was not always returning a list in `matched_processes` (could be $null, or a single hash table)

jborean93 commented:
@dagwieers I think I've found the cause of the issues with the async task. I've enabled logging and had a look at the module invocation output and found the following error message

```
fatal: [server2016.domain.local]: FAILED! => {
    "ansible_async_watchdog_pid": 6856,
    "ansible_job_id": "536774098749.7864",
    "attempts": 1,
    "changed": false,
    "finished": 1,
    "msg": "failed to parse module output: A hash table can only be added to another hash table.",
    "results_file": "C:\\Users\\vagrant\\AppData\\Local\\Temp\\.ansible_async\\536774098749.7864",
    "started": 1,
    "stderr": "Exception calling \"Invoke\" with \"0\" argument(s): \"The running command stopped because the \r\npreference variable \"ErrorActionPreference\" or common parameter is set to Stop: Illegal operation \r\nattempted on a registry key that has been marked for deletion. \"\r\nAt line:40 char:5\r\n+     $o
utput = $ps.Invoke()\r\n+     ~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException\r\n    + FullyQualifiedErrorId : ActionPreferenceStopException\r\n \r\n",                                                                                                            
    "stderr_lines": [
        "Exception calling \"Invoke\" with \"0\" argument(s): \"The running command stopped because the ",
        "preference variable \"ErrorActionPreference\" or common parameter is set to Stop: Illegal operation ",
        "attempted on a registry key that has been marked for deletion. \"",
        "At line:40 char:5",
        "+     $output = $ps.Invoke()",
        "+     ~~~~~~~~~~~~~~~~~~~~~~",
        "    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException",
        "    + FullyQualifiedErrorId : ActionPreferenceStopException",
        " "
    ],
    "stdout": "",
    "stdout_lines": []
}
```

I am able to easily replicate this by running the current tests you have by running as a user that is not current logged on locally which tells me that once the async process starts and the WinRM process exists it unloads the user's profile causing the win_shell task to fail. There may be some successful runs depending on the time it takes for the initial WinRM process to exit.

As for a fix, I'm not sure yet, I need to look into it a bit further but for now you may just need to disable the win_shell async task until we get it fixed.

jborean93 commented:
No idea why it's happening but I am having more luck using Get-Process over WMI. I highly recommend you change the code in `Get-ProcessMatchesFilter` to use that. Basically to get a list of all processes run `Get-Process`, you can use the `Id` and `Name` properties to match `ProcessName` and `ProcessId`. To get the current username you then need to run `(Get-Process -Id $pid -IncludeUsername).UserName` which returns a string. The string format is different but you can use https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.SID.psm1 to convert both the input `owner` option as well as the output from the `Get-Process` cmdlet. From there it's just a matter of comparing the SID strings which should either be identical or not, no need to compare strings and expect the user to conform to a particular format for the owner option.

dagwieers commented:
@jborean93 In any case it would be nice if the real issue would end up in the output to the user. Or we will get more cases like this one ;-) But at least we understand what is going on now, and how async can influence the normal flow of a running process (if I understand correctly).

jborean93 commented:
@dagwieers that's what https://github.com/ansible/ansible/pull/44807 is for. It's not perfect and I had to do some more to get that stacktrace and better error message but it still contains the error message we normally get which helps. It's weird that the `Get-CimInstance Win32_Process` is reliant on the users profile but weirder things happen.

dagwieers commented:
@jborean93 I rewrote the function, and first used `-IncludeUserName` which worked, but required elevated privileges. So I am kinda stuck now :-/

dagwieers committed:
Various improvements

This PR includes:
- Use Get-Process instead of CIM Win32_Process
- Rewrite of process filter logic (speedup)
- Fix error messages
- Fixes to documentation, examples and return output

crossan007 commented on lib/ansible/modules/windows/win_wait_for_process.ps1:
```diff
@@ -64,99 +62,109 @@ Function Get-ProcessMatchesFilter {
-    $CIMProcesses = Get-CimInstance Win32_Process
-    foreach ($CIMProcess in $CIMProcesses)
+
+    $WMIProcesses = Get-Process -IncludeUserName
```
Huh.  Not sure how I missed `-IncludeUserName` and deviated to the WMI calls...

dagwieers replied:
Well, it appears `-IncludeUserName` only works from 2012R2 up ?

crossan007 replied:
Ah.  That's probably a good reason.

Pretty sure I was reading this:  https://social.technet.microsoft.com/Forums/lync/en-US/45c541e0-fb49-4ac7-bcaf-738315a70863/getprocess-and-username?forum=winserverpowershell

crossan007 replied:
Perusing the interwebs, it seems the `-IncludeUserName` is PoSH4 and up + Admin, as you've described

crossan007 replied:
https://thinkpowershell.com/get-process-name-and-owner-user-name/

dagwieers committed:
win_wait_for_process: Limit to PowerShell 4 and higher

dagwieers commented:
Re-testing with https://github.com/ansible/ansible/pull/44807 enabled as I cannot reproduce this locally ATM.

dagwieers commented:
Also beware that the module from now is PS4+ only, we can fix that later, if needed.

jborean93 commented on lib/ansible/modules/windows/win_wait_for_process.ps1:
```diff
@@ -64,99 +64,108 @@ Function Get-ProcessMatchesFilter {
-            else {
-                $include = $include -and ($ProcessNameExact -eq $CIMProcess.ProcessName)
+
+    $WMIProcesses = Get-Process -IncludeUserName
```
To keep compatibility with PSv3 you should optionally include the `-IncludeUserName` parameter depends on if `$Owner` is set. This way you can say in the docs, the `owner` option requires PSv4+ but the module will still work with other ones if you are on that version.

```
$parameters = @{}
if ($Owner) {
    $parameter.IncludeUserName = $true
}
$WMIProcesses = Get-Process @parameters
```

Also WMIProcesses isn't a good name, this get's all processes and not WMI ones.

dagwieers replied:
It used to be called $CIMProcesses, that's why I named it $WMIProcesses. But yeah, could as well be $Processes.

jborean93 commented on lib/ansible/modules/windows/win_wait_for_process.ps1:
```diff
@@ -64,99 +64,108 @@ Function Get-ProcessMatchesFilter {
+    foreach ($WMIProcess in $WMIProcesses) {
+
+        # If a process name was specified in the filter, validate that here.
+        if (-not [String]::IsNullOrEmpty($ProcessNamePattern)) {
```
You don't have to change it but this can be simplified with `if ($ProcessNamePattern) {`.

dagwieers replied:
Yup, I prefer to change that ;-)

dagwieers committed:
Improve RESULT documentation

dagwieers committed:
Last minute fixes for CI

dagwieers committed:
Catch Powershell exceptions

dagwieers committed:
Increase timeout to make tests more stable

jborean93 commented:
@dagwieers in case you are still having issues tomorrow with this and need a dummy service. I've created a service which just logs each action to the event log. The binary can be downloaded from https://s3.amazonaws.com/ansible-ci-files/test/integration/roles/test_win_service/SleepService.exe and can be installed with

```
- name: install dummy service
  win_service:
    name: dummy service
    path: C:\path\to\SleepService.exe
    state: started
```

From there you can decide how many you want to run as well as what account runs it.

dagwieers commented:
@jborean93 I think this is ready to be merged as-is. Finally !

jborean93 commented on lib/ansible/modules/windows/win_wait_for_process.py:
```diff
@@ -66,34 +89,43 @@
-  type: int
+  type: complex
+  contains:
+    Id:
```
I don't like the fact these aren't snake_case, surely we can convert

dagwieers replied:
Hints to do this efficiently in Powershell are welcome :)

jborean93 replied:
In this case I would just manually do it, especially turning `Id` to `pid` to match the option names. Otherwise there is https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/powershell/Ansible.ModuleUtils.CamelConversion.psm1.

dagwieers replied:
I fixed this. Please re-review :-)

jborean93 commented on lib/ansible/modules/windows/win_wait_for_process.ps1:
```diff
@@ -1,59 +1,62 @@
 $ErrorActionPreference = "Stop"
 
+trap {
+    $result.exception = "$($_ | Out-String)`r`n$($_.ScriptStackTrace)"
```
Maybe add a comment as to why the trap is here.

dagwieers replied:
Should we keep it in ?

jborean93 replied:
up to you, I haven't fully thought out what some of the consequences could be but I was ultimately planning on having the exec_wrapper inject something like this into the module on execution to give us a prettier output.

dagwieers replied:
Yes please ! I think it will help us with some behaviour.
The sooner the better if you ask me ;-)

dagwieers replied:
Added note !

jborean93 reviewed (commented):
few minor nits
